### PR TITLE
Implement WPT RFC 72

### PR DESCRIPTION
**Pull request by letitz**
This PR defines 4 new ports:

 - `http-private`
 - `http-public`
 - `https-private`
 - `https-public`

These are used to run additional HTTP and HTTPS servers on `127.0.0.1`. These new (IP address, port) pairs are configured such that their IP address space in Chrome is overridden to `private` and `public` as indicated in their names. Other browsers lack the ability to configure such overrides, so they are omitted in this PR.

This allows testing the [Private Network Access](https://wicg.github.io/private-network-access/) specification more extensively than what is already in `fetch/private-network-access/`. The command-line switch has yet to land in Chromium (see https://crrev.com/c/2851238) so no tests can yet be written, but I am interested in this change getting reviewed anyway. If need be, I can wait to submit this until such tests are available, or I can write tests that expect failure.

This PR is destined to supplant #28768, which was the first ill-fated attempt at implementing RFC 72, before https://github.com/web-platform-tests/rfcs/pull/83 came along.

### CODE REVIEW

**Comment by letitz** [comment]
Not sure why the "Azure Pipelines (tools/ unittests: macOS + Python 3.6)" check failed. I cannot see the logs for the step that purportedly failed: "Install Python packages". Closing and reopening so as to trigger CI again.

**Comment by letitz** [comment]
OK CI is happy now. @foolip FYI.

**Comment by letitz** [comment]
@LukeZielinski friendly ping. I'd like to start writing full WPTs for Private Network Access based on this change before Chrome 92 ships to stable.

**Review by jgraham — COMMENTED** [review]
I don't know if it would have been possible to reuse the existing servers for "public"? If the states are either "private" or "public" I would have preferred to just have one of each to cut down startup overhead.

If that's not possible, this looks fine.

**Comment by foolip** [comment]
@letitz I'll go ahead and review this for you.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (foolip) [inline tools/wptrunner/wptrunner/browsers/chrome.py:100]: Is this same argument needed to test this on Edge (edgechromium.py) as well? How about Chrome for Android?

**Review by foolip — COMMENTED** [review]
Can this be tested? Not sure if unit tests are useful/practical, but at least one test under infrastructure/ that will prevent this from regressing would be great.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (letitz) [reply tools/wptrunner/wptrunner/browsers/chrome.py:100]: Great question, I forgot since I wrote the command-line argument code. I'll need to check whether this is available to all embedders or just desktop Chrome for the moment.

**Comment by letitz** [comment]
Thanks @jgraham, @foolip!

@jgraham: There are 3 possible states: `public`, `private` and `local`. The existing `http` and `https` fulfill the role of `http-local` and `https-local` because they are served from loopback, hence are classified in the `local` address space by default.

@foolip: Within `infrastructure/` we can test that the servers are running correctly when configured so, and I guess we can try to test command-line arguments. I'm not sure we should be spinning up browsers and exercising their behavior, I think that belongs in `fetch/private-network-access/`, which should prevent functional regressions?

**Comment by foolip** [comment]
@letitz the tests in `infrastructure/` have expectations checked in and are run when we change code in tools/, so that prevents accidental regressions. Any other tests, like those in fetch/, could be broken without it blocking our CI.

Since part of this support is in the browser setup (command line argument) I think such an end-to-end test is best, but you might be able to test it with wptserve+wptrunner unit tests, if you'd prefer.

**Comment by letitz** [comment]
Got it. I'll look into writing an end-to-end test then, and fall back to some kind of unit test if that fails. Do you have pointers to other end-to-end tests within `infrastructure/` from which I could draw inspiration?

**Comment by letitz** [comment]
Well, I found `infrastructure/assumptions/` which seems appropriate. Thinking about how to test this, however, makes me think this is ill-suited.

The Private Network Access spec used to specify a `window.addressSpace` property that could be used to test this fairly easily. However upon discussing it with Chromium privacy folks and the TAG, I found that it was somewhat harmful so I removed it (wicg/private-network-access#21).

The only way we can test this now is to check that a non-secure `public` page cannot make subresource requests to `private` or `local` servers. These are the tests I am planning to write in `fetch/private-network-access/`. It feels redundant to copy them (or one of them at least) in `infrastructure/`, but maybe that's fine. More importantly, this test will fail on all browsers but Chrome, which makes me think it is inappropriate to run CI against. WDYT?

**Comment by jgraham** [comment]
The CI thing isn't problematic; we can mark those as `expected: FAIL` in other browsers. I kind of agree that the tests are going to be a bit redundant with the actual feature tests. At the very least we should check that the servers are able to respond to requests, even if we don't also check the correct configuration.

**Comment by letitz** [comment]
Added tests to check that the other servers nominally serve files from the WPT tree. I can add a test checking that fetches are blocked correctly, as a regression test for the command-line config part. How do I mark a test as expected to fail?

**Comment by letitz** [comment]
CI passed on the basic tests. Will push integration test and run CI on it - I expect it to fail on non-Chrome browsers.

**Comment by letitz** [comment]
@jgraham How do I mark a test as expected to fail for certain browsers?

@foolip I believe Chrome on Android and Edge on Chromium will support the same command-line flag, since all the logic sits at the `content/` layer or below. Just in case, I'd like to defer that to a followup PR if that's OK with you. My thinking is that I'll first write a couple tests in `fetch/private-network-access/` via the Chromium repository, then I'll draft a PR to add the flag to both Chrome on Android and Edge on Chromium and see what CI says.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (letitz) [reply tools/wptrunner/wptrunner/browsers/chrome.py:101]: I believe Chrome on Android and Edge on Chromium will support the same command-line flag, since all the logic sits at the `content/` layer or below. Just in case, I'd like to defer that to a followup PR if that's OK with you. My thinking is that I'll first write a couple tests in `fetch/private-network-access/` via the Chromium repository, then I'll draft a PR to add the flag to both Chrome on Android and Edge on Chromium and see what CI says.

**Comment by jgraham** [comment]
Create a file like `infrastructure/metadata/infrastructure/assumptions/non-local-ports.sub.window.html.ini` there are "ini" files (not really ini files) which you can use like

```
[non-local-ports.sub.window.html]
  [Fetch from http-public to local http fails.]
     expected:
      if product != "chrome": FAIL
```

Then to test it's correct do something like `./wpt run --metadata infrastructure/metadata/ firefox infrastructure/assumptions/non-local-ports.sub.window.html`

**Comment by jgraham** [comment]
I didn't test that btw so I might have got it wrong :)

**Comment by letitz** [comment]
Got it, thanks!

**Comment by letitz** [comment]
Not sure why the Community-TC tests failed even though the Azure `infrastructure/` tests passed... I can see the test using `--metadata infrastructure/metadata` as it should.

**Comment by letitz** [comment]
@jgraham It seems the expectation is not respected by CI on Community-TC. Is this a known issue?

**Comment by letitz** [comment]
Ah, it seems the test is failing as expected on Firefox, but also unexpectedly on Chrome dev 91.0.4472.19. I was expecting it to use Chrome 92, which passes the test.

How can we update the version of Chrome used by `infrastructure/` tests running on Community-TC?

If that's not possible, I can split the fetch test into its own followup PR, land the passing tests and wait until Community-TC picks up Chrome 92 to land the fetch test.

**Comment by letitz** [comment]
Split out the fetch test from this PR, so we can land this with passing tests.

@foolip @jgraham PTAL, thanks!

File: infrastructure/assumptions/non-local-ports.sub.window.js
Comment (foolip) [inline infrastructure/assumptions/non-local-ports.sub.window.js:38]: Can you `const res = await fetch(...)` and assert `res.ok`? If not, some errors might be hidden here, right?

File: infrastructure/assumptions/non-local-ports.sub.window.js
Comment (foolip) [inline infrastructure/assumptions/non-local-ports.sub.window.js:34]:
```suggestion
  const url = resolveUrl("/common/blank.html", {
```

Since the content isn't used, might as well use an existing empty file.

File: tools/wptrunner/wptrunner/browsers/chrome.py
Comment (foolip) [reply tools/wptrunner/wptrunner/browsers/chrome.py:101]: Follow-up is OK :)

**Review by foolip — COMMENTED** [review]
I'm happy with this now, just a few nits I'm sure will be easy to resolve.

@jgraham do you want to take a final look too?

File: infrastructure/assumptions/non-local-ports.sub.window.js
Comment (letitz) [reply infrastructure/assumptions/non-local-ports.sub.window.js:38]: You're right. Fixing.

File: infrastructure/assumptions/non-local-ports.sub.window.js
Comment (letitz) [reply infrastructure/assumptions/non-local-ports.sub.window.js:34]: Good idea, thanks!

Comment (letitz) [reply infrastructure/assumptions/non-local-ports.sub.window.js:34]: Actually I needed to add `Access-Control-Allow-Origin: *` to allow accessing `response.ok` on the cross-origin response, so I kept the `resources/exists.txt` file.

**Comment by letitz** [comment]
Weird errors during CI. I'll rebase and run again.

**Comment by letitz** [comment]
It seems that test is flaky, since it now passed. Removing the extra logging.

**Comment by letitz** [comment]
Hah! Now that I've removed the extra logging, the tests pass again :upside_down_face:. Re-adding logging.

It seems this PR might be making this test extra flaky. Could it be because it increases the number of bound ports?

**Comment by letitz** [comment]
And now it passes. I'm getting a feeling the CI infrastructure is mocking me! :fearful: 

Closing and re-opening to run tests again.

**Comment by letitz** [comment]
Cannot get it to fail with extra logging. Removing once more...

**Comment by jgraham** [comment]
Ah, so we have [code](We're supposed to wait on startup for all the servers to https://searchfox.org/mozilla-central/source/testing/web-platform/tests/tools/wptrunner/wptrunner/environment.py#244-283) to check all the servers start in wptrunner. But maybe the resources/ tests don't have anything like that.

**Comment by letitz** [comment]
I thought I had seen something like that somewhere, and was looking around for it. Thanks for the pointer!

**Comment by jgraham** [comment]
https://github.com/web-platform-tests/wpt/pull/29344 refactors the wait-for-servers code here in a way that should make it more reliable. Maybe try rebasing onto that branch and see if it fixes the problem?

**Comment by letitz** [comment]
Oh, that's neat! I added a small change to the server start logic for `resources/test/` that checks that the HTTPS server is live as well as the HTTP server before running tests. I think that should be enough for now, let's see what CI says.

**Comment by letitz** [comment]
OK! Tests passed :tada:. Can I merge this for now, and #29344 can provide the proper fix on top?

**Comment by jgraham** [comment]
Yes, happy for you to merge.

File: infrastructure/assumptions/resources/exists.txt
Comment (jgraham) [inline infrastructure/assumptions/resources/exists.txt:1]: I might have put this in `/common`, or even just in `/infrastructure`

**Review by jgraham — APPROVED** [review]

File: infrastructure/assumptions/resources/exists.txt
Comment (letitz) [reply infrastructure/assumptions/resources/exists.txt:1]: I can move it.

**Comment by letitz** [comment]
Thanks @jgraham :tada:

**Pull request merged**
